### PR TITLE
[SYCL][Doc] Add improved guidelines for runtime test development

### PR DESCRIPTION
**Pull request by alexbatashev**
New guidelines should provide clear expectations from incoming pull requests and also give hints on how to write tests for particular types of changes.

### CODE REVIEW

**Comment by alexbatashev** [comment]
Some improvements for PI Mock framework from these BKMs are not merged yet: https://github.com/intel/llvm/pull/5358

File: sycl/doc/dev/RuntimeTestingBKMs.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
+# SYCL Runtime and Device Headers Testing BKMs
```
Comment (pvchupin) [inline sycl/doc/dev/RuntimeTestingBKMs.md:1]:
```suggestion
# SYCL Runtime and Device Headers Testing Guidelines
```

Comment (pvchupin) [reply sycl/doc/dev/RuntimeTestingBKMs.md:1]: Please rename md filename/refs accordingly.

File: sycl/doc/dev/RuntimeTestingGuidelines.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+### Device Image Mock
+
+SYCL Runtime discoveres user-defined kernels by looking up kernel names in
+registered device images. These images contain all information about SPIR-V or
+binary device code, including kernel names and some metadata passed from the
+compiler frontend.
+
+`helpers/PIImage.hpp` header provides routines to insert a fake device image
+into Program Manager, and emulate precense of particular kernels.
```
Comment (steffenlarsen) [inline sycl/doc/dev/RuntimeTestingGuidelines.md:51]:
```suggestion
into Program Manager, and emulate presence of particular kernels.
```

File: sycl/doc/dev/RuntimeTestingGuidelines.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
+# SYCL Runtime and Device Headers Testing BKMs
```
Comment (alexbatashev) [inline sycl/doc/dev/RuntimeTestingGuidelines.md:1]:
```suggestion
# SYCL Runtime and Device Headers Testing Guidelines
```

File: sycl/doc/dev/RuntimeTestingGuidelines.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+## SYCL Host Runtime Testing
+
+Most changes to the SYCL Host Runtime are device-agnostic, thus can and must be
+tested in isolation from the rest of the SYCL toolchain. Below is the
+information on some techniques, that allow one to isolate these parts of the
+runtime and simulate different behavior of low-level runtimes.
+
+### Plugin Interface Mock
+
+PI Mock layer provides capabilities to override plugins' behavior. There're two
```
[the remark this answers is not in the document]
Comment (steffenlarsen) [reply sycl/doc/dev/RuntimeTestingGuidelines.md:12]: Should be find to just expand them, i.e. "there're" ->"there are" and "there's"->"there is".

**Comment by pvchupin** [comment]
It looks good to me but would be good to see more reviews/comments from @intel/llvm-reviewers-runtime who write the tests and expected to be compliant with this doc.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -104,12 +104,20 @@ end-to-end or SYCL-CTS tests.
...
+    - PRs changing the way PI APIs are called: provide a unit test for new behavior,
+      see [Guidelines](sycl/doc/dev/RuntimeTestingGuidelines.md#Plugin-interface-mock)
+    - PRs changing graph structure or nodes scheduling: provide a unit test for new
+      behavior, see [Guidelines](sycl/doc/dev/RuntimeTestingGuidelines.md#SYCL-host-runtime-testing)
+    - If PR adds a SYCL API, that is required by the specification to throw
+      exceptions: provide a unit test for error path,
+      see [Guidelines](sycl/doc/dev/RuntimeTestingGuidelines.md#Plugin-interface-mock)
+    - PRs using "magic" builtins in device code: provide a LIT test for LLVM IR,
+      generated with new code, see [Guidelines](sycl/doc/dev/RuntimeTestingGuidelines.md#sycl-device-headers)
+    - PRs adding new data types as part of ABI: provide a LIT test for new data type
```
Comment (romanovvlad) [inline CONTRIBUTING.md:116]:
```suggestion
    - PRs adding new user accessible data types or making some type a part of SYCL library ABI: provide a LIT test for new data type
```

File: sycl/doc/dev/RuntimeTestingGuidelines.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+functions accept a function pointer or a lambda. These overrides can be used
+to check, that a particular function has been called, or to emulate behavior
+of a low-level runtime.
+
+> If a pull request changes the way runtime plugin calls PI functions, it must
+> also contain a unit test with PI mock to verify these changes.
+
+> If a pull request adds new API, that may throw an exception, it must also
+> contain a unit test with exception path covered. In case exception depends
+> on low-level runtime call result, use PI Mock to simulate faulty call.
```
Comment (romanovvlad) [inline sycl/doc/dev/RuntimeTestingGuidelines.md:41]: Is it planned section? Looks like a duplication of what is stated in `CONTRIBUTING.md`.

**Review by romanovvlad — COMMENTED** [review]
LGTM

File: sycl/doc/dev/RuntimeTestingGuidelines.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+## SYCL Host Runtime Testing
+
+Most changes to the SYCL Host Runtime are device-agnostic, thus can and must be
+tested in isolation from the rest of the SYCL toolchain. Below is the
+information on some techniques, that allow one to isolate these parts of the
+runtime and simulate different behavior of low-level runtimes.
+
+### Plugin Interface Mock
+
+PI Mock layer provides capabilities to override plugins' behavior. There are two
```
Comment (vladimirlaz) [inline sycl/doc/dev/RuntimeTestingGuidelines.md:12]: There are 2 options here. Could you add a recommendation on what path to go in different cases (if there is any difference)?

File: sycl/doc/dev/RuntimeTestingGuidelines.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+SYCL Runtime discoveres user-defined kernels by looking up kernel names in
+registered device images. These images contain all information about SPIR-V or
+binary device code, including kernel names and some metadata passed from the
+compiler frontend.
+
+`helpers/PIImage.hpp` header provides routines to insert a fake device image
+into Program Manager, and emulate presence of particular kernels.
+
+In the default compiler flow, Clang generates a so-called integration header,
```
Comment (vladimirlaz) [inline sycl/doc/dev/RuntimeTestingGuidelines.md:53]: do we have any recommendations for the device footer?

File: sycl/doc/dev/RuntimeTestingGuidelines.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+the correctness of compiler frontend output in the form of LLVM IR (i.e.
+check calls to particular intrinsics exist, function mangling is correct, etc).
+
+Find examples of such tests in `sycl/test/check_device_code` directory.
+
+Refer to [FileCheck](https://llvm.org/docs/CommandGuide/FileCheck.html) and
+[LIT](https://llvm.org/docs/CommandGuide/lit.html) tools documentation for
+more info.
+
+### Changing ABI
```
Comment (vladimirlaz) [inline sycl/doc/dev/RuntimeTestingGuidelines.md:99]: Mostly duplication of lines 73-83

File: sycl/doc/dev/RuntimeTestingGuidelines.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+Whenever a new data structure is added as part of ABI (either between device and
+host, or between two device images), a test for data layout must be added,
+same as for host runtime. If the data structure can be used both on host and
+device sides of user code, both tests must be present.
+
+Refer to [ABI Policy Guide](https://intel.github.io/llvm-docs/ABIPolicyGuide.html)
+for more info.
+
+## End-to-end tests
```
Comment (vladimirlaz) [inline sycl/doc/dev/RuntimeTestingGuidelines.md:109]: Do we want to check expected warnings/errors in  https://github.com/intel/llvm/tree/sycl/sycl/test/warnings?

**Comment by bader** [comment]
@alexbatashev, please, address code review comments and merge conflicts.

**Review by bader — COMMENTED** [review]
@alexbatashev, please, address comments and resolve merge conflicts.

**Pull request closed without merging**